File: Algoritimo-Estrutruras-de-Dados/Aeds-ll/2025_01/trabalhos_praticos/TP_04/questao_03/Show.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

int count = 0;
float tempo = 0.0;
clock_t inicio;
clock_t fim;

char toLower(char c){
    char resp = c;
    if(c >= 'A' && c <= 'Z'){
        resp = c + 32;
    }
    return resp;
}

int IsFim(char line[]){
    return !(line[0] == 'F' && line[1] == 'I' && line[2] == 'M' && line[3] == '\0');
}

int Strleng(char str[]){
    int length = 0;
    while(str[length] != '\0'){
        length++;
    }
    return length;
}

void Strcopy(char dest[], char src[]){
    int i = 0;
    while(src[i] != '\0'){
        dest[i] = src[i];
        i++;
    }
    dest[i] = '\0';
}

int isEqual(char str1[], char str2[]){
    int i = 0;
    int resp = 1;
    int len1 = Strleng(str1);
    int len2 = Strleng(str2);
    if(len1 == len2){
        while(i < len1 && resp){
            if(str1[i] != str2[i]){
                resp = 0;
            }
            i++;
        }
    } else{
        resp = 0;
    }
    return resp;
}

typedef struct{
    int ano;
    char mes[20];
    int dia;
} Data;

void Init_Date(Data *data){
    data->ano = 0;
    Strcopy(data->mes, "NaN");
    data->dia = 0;
}

void parseDate(Data *data, const char *source){
    if(source == NULL || source[0] == '\0'){
        Strcopy(data->mes, "March");
        data->dia = 1;
        data->ano = 1900;
    } else{
        char aux[100] = "";
        int auxIndex = 0;
        int step = 0;
        int i = 0;
        while(source[i] != '\0'){
            if(source[i] != ' ' && source[i] != ','){
                aux[auxIndex++] = source[i];
                aux[auxIndex] = '\0';
            } else{
                if(step == 0){
                    Strcopy(data->mes, aux);
                } else if(step == 1){
                    data->dia = atoi(aux);
                }
                auxIndex = 0;
                aux[0] = '\0';
                step++;
            }
            i++;
        }
        if(!isEqual(aux, "")){
            data->ano = atoi(aux);
        }
    }
}

typedef struct{
    char show_id[20];
    char type[20];
    char title[200];
    char director[100];
    char **cast;
    int cast_count;     
    char country[100];
    Data data_added;
    int release_year;
    char rating[20];
    char duration[20];
    char **listed_in;
    int listed_count;
} Show;

typedef struct No{
    Show elemento;
    struct No *esq, *dir;
} No;

typedef struct ArvoreAVL{
    No *raiz;
} ArvoreAVL;

ArvoreAVL *arvoreConstrutor(){
    ArvoreAVL *i = (ArvoreAVL *)malloc(sizeof(ArvoreAVL));
    i->raiz = NULL;
    return i;
}

No *construtorNo(Show show){
    No *i = (No *)malloc(sizeof(No));
    i->elemento = show;
    i->dir = i->esq = NULL;
    return i;
}

int getAltura(No *i){
    if(i == NULL){
        return -1;
    }
    int alturaEsq = getAltura(i->esq) + 1;
    int alturaDir = getAltura(i->dir) + 1;
    return (alturaEsq > alturaDir ? alturaEsq : alturaDir);
}

int getFator(No *i){
    int alturaEsq = getAltura(i->esq) + 1;
    int alturaDir = getAltura(i->dir) + 1;
    return (alturaDir - alturaEsq);
}

No *rotacaoDir(No *i){
    No *tmp = i->esq;
    i->esq = tmp->dir;
    tmp->dir = i;
    return tmp;
}

No *rotacaoEsq(No *i){
    No *tmp = i->dir;
    i->dir = tmp->esq;
    tmp->esq = i;
    return tmp;
}

No *inserirRec(No *i, Show show){
    if(i == NULL){
        i = construtorNo(show);
    } else if(strcmp(show.title, i->elemento.title) > 0){
        i->dir = inserirRec(i->dir, show);
    } else if(strcmp(show.title, i->elemento.title) < 0){
        i->esq = inserirRec(i->esq, show);
    } else{
        printf("ERRO, VALOR INVALIDO");
    }

    if(getFator(i) == -2){
        if(getFator(i->esq) == 1){
            i->esq = rotacaoEsq(i->esq);
        }
        i = rotacaoDir(i);
    } else if(getFator(i) == 2){
        if(getFator(i->dir) == -1){
            i->dir = rotacaoDir(i->dir);
        }
        i = rotacaoEsq(i);
    }

    return i;
}

void inserir(ArvoreAVL *arvore, Show show){
    arvore->raiz = inserirRec(arvore->raiz, show);
}

void pesquisa(No *i, char title[]){
    if(i == NULL){
        printf(" NAO\n");
    } else if(strcmp(title, i->elemento.title) < 0){
        printf(" esq");
        count++;
        pesquisa(i->esq, title);
    } else if(strcmp(title, i->elemento.title) > 0){
        printf(" dir");
        count++;
        pesquisa(i->dir, title);
    } else{
        printf(" SIM\n");
        count++;
    }
}

void pesquisar(ArvoreAVL *arvore, char title[]){
    printf("raiz");
    pesquisa(arvore->raiz, title);
}

void Show_init(Show *s){
    Strcopy(s->show_id, "NaN");
    Strcopy(s->type, "NaN");
    Strcopy(s->title, "NaN");
    Strcopy(s->director, "NaN");

    s->cast = (char **)malloc(sizeof(char *));
    s->cast[0] = (char *)malloc(20 * sizeof(char));
    Strcopy(s->cast[0], "NaN");
    s->cast_count = 1;

    Strcopy(s->country, "NaN");

    Init_Date(&s->data_added);

    s->release_year = 0;
    Strcopy(s->rating, "NaN");
    Strcopy(s->duration, "NaN");


    s->listed_in = (char **)malloc(sizeof(char *));
    s->listed_in[0] = (char *)malloc(20 * sizeof(char));
    Strcopy(s->listed_in[0], "NaN");
    s->listed_count = 1;
}

int compareStrings(char str1[], char str2[]){
    int i = 0;
    int result = 0;

    while(str1[i] != '\0' && str2[i] != '\0' && result == 0){
        char c1 = toLower(str1[i]);
        char c2 = toLower(str2[i]);
        if(c1 != c2){
            result = c1 -  c2;
        }
        i++;
    }

    if(result == 0){
        result = Strleng(str1) - Strleng(str2);
    }
    return result;
}

void OrdenarCast(char **cast, int size){
    int i, j;
    char temp[100];

    for(i = 0; i < size; i++){
        for(j = 0; j < size - 1; j++){
            if(compareStrings(cast[j], cast[j + 1]) > 0){
                Strcopy(temp, cast[j]);
                Strcopy(cast[j], cast[j + 1]);
                Strcopy(cast[j + 1], temp);
            }
        }
    }
}

void Show_Imprimir(Show *s){
    printf("=> %s ## %s ## %s ## %s ## [",
           s->show_id, s->title, s->type, s->director);

    for(int i = 0; i < s->cast_count; i++){
        printf("%s", s->cast[i]);
        if(i < s->cast_count - 1){
            printf(", ");
        }
    }

    printf("] ## %s ## %s %d, %d ## %d ## %s ## %s ## [",
           s->country, s->data_added.mes, s->data_added.dia,
           s->data_added.ano, s->release_year, s->rating, s->duration);

    for(int i = 0; i < s->listed_count; i++){
        printf("%s", s->listed_in[i]);
        if(i < s->listed_count - 1){
            printf(", ");
        }
    }

    printf("] ##\n");
}

void Show_Ler(Show *show, char* line){
    char temp[11][1000];
    int TMPindex[11] = {0};
    int i = 0;
    int campo = 0;

    while(line[i] != '\0' && campo < 11){
        if(line[i] == '"'){
            i++;
            while(line[i] != '"' || (line[i] == '"' && line[i + 1] == '"')){
                if(line[i] == '"' && line[i + 1] == '"'){
                    i += 2;
                } else{
                    temp[campo][TMPindex[campo]++] = line[i++];
                }
            }
            i++;
            if (line[i] == ',')
                i++;
            temp[campo][TMPindex[campo]] = '\0';
            campo++;
        } else{
            while(line[i] != ',' && line[i] != '\0'){
                temp[campo][TMPindex[campo]++] = line[i++];
            }
            if(line[i] == ','){
                i++;
            }
            temp[campo][TMPindex[campo]] = '\0';
            campo++;
        }
    }

    Strcopy(show->show_id, temp[0][0] ? temp[0] : "NaN");
    Strcopy(show->type, temp[1][0] ? temp[1] : "NaN");
    Strcopy(show->title, temp[2][0] ? temp[2] : "NaN");
    Strcopy(show->director, temp[3][0] ? temp[3] : "NaN");
    Strcopy(show->country, temp[5][0] ? temp[5] : "NaN");
    parseDate(&show->data_added, temp[6]);
    show->release_year = atoi(temp[7]);
    Strcopy(show->rating, temp[8][0] ? temp[8] : "NaN");
    Strcopy(show->duration, temp[9][0] ? temp[9] : "NaN");

    if(temp[4][0]){
        show->cast_count = 1;
        for(i = 0; temp[4][i]; i++){
            if(temp[4][i] == ','){
                show->cast_count++;
            }
        }

        show->cast = (char **)malloc(show->cast_count * sizeof(char *));
        for (i = 0; i < show->cast_count; i++)
        {
            show->cast[i] = (char *)malloc(100 * sizeof(char));
        }

        int IndexC = 0;
        int currentCast = 0;
        i = 0;
        while(temp[4][i]){
            if(temp[4][i] == ','){
                show->cast[currentCast][IndexC] = '\0';
                currentCast++;
                IndexC = 0;
                i++;
                while(temp[4][i] == ' '){
                    i++;
                }
            } else{
                show->cast[currentCast][IndexC++] = temp[4][i++];
            }
        }
        show->cast[currentCast][IndexC] = '\0';
        OrdenarCast(show->cast, show->cast_count);
    } else{
        show->cast_count = 1;
        show->cast = (char **)malloc(sizeof(char *));
        show->cast[0] = (char *)malloc(20 * sizeof(char));
        Strcopy(show->cast[0], "NaN");
    }

    if(temp[10][0]){
        show->listed_count = 1;
        for(i = 0; temp[10][i]; i++){
            if(temp[10][i] == ','){
                show->listed_count++;
            }
        }

        show->listed_in = (char **)malloc(show->listed_count * sizeof(char *));
        for(i = 0; i < show->listed_count; i++){
            show->listed_in[i] = (char *)malloc(100 * sizeof(char));
        }

        int IndexL = 0;
        int currentList = 0;
        i = 0;
        while (temp[10][i]){
            if(temp[10][i] == ','){
                show->listed_in[currentList][IndexL] = '\0';
                currentList++;
                IndexL = 0;
                i++;
                while(temp[10][i] == ' '){
                    i++;
                }
            } else{
                show->listed_in[currentList][IndexL++] = temp[10][i++];
            }
        }
        show->listed_in[currentList][IndexL] = '\0';
    } else{
        show->listed_count = 1;
        show->listed_in = (char **)malloc(sizeof(char *));
        show->listed_in[0] = (char *)malloc(20 * sizeof(char));
        Strcopy(show->listed_in[0], "NaN");
    }
}

void copyCast(Show *copy, Show *source){
    copy->cast_count = source->cast_count;
    copy->cast = (char**)malloc(copy->cast_count * sizeof(char*));
    
    for(int i = 0; i < copy->cast_count; i++){
        copy->cast[i] = (char*)malloc(100 * sizeof(char));
        Strcopy(copy->cast[i], source->cast[i]);
    }
}

void copyListed(Show *copy, Show *source){
    copy->listed_count = source->listed_count;
    copy->listed_in = (char**)malloc(copy->listed_count * sizeof(char*));
    
    for(int i = 0; i < copy->listed_count; i++){
        copy->listed_in[i] = (char*)malloc(100 * sizeof(char));
        Strcopy(copy->listed_in[i], source->listed_in[i]);
    }
}

void clone(Show *array, Show *source){
    Strcopy(array->show_id, source->show_id);
    Strcopy(array->type, source->type);
    Strcopy(array->title, source->title);
    Strcopy(array->director, source->director);
    Strcopy(array->country, source->country);
    array->release_year = source->release_year;
    Strcopy(array->rating, source->rating);
    Strcopy(array->duration, source->duration);

    array->data_added.ano = source->data_added.ano;
    array->data_added.dia = source->data_added.dia;
    Strcopy(array->data_added.mes, source->data_added.mes);

    copyCast(array,source);
    copyListed(array,source);
}

int Pesquisar_ID(Show shows[], int size, char* id){
    int resp = 0;
    int i = 0;
    while(i < size){
        if(isEqual(shows[i].show_id, id)){
            resp = i;
            i = size;
        }
        i++;
    }
    return resp;
}

void FreeShow(Show *s){
    for(int i = 0; i < s->cast_count; i++){
        free(s->cast[i]);
    }
    free(s->cast);
    for(int i = 0; i < s->listed_count; i++){
        free(s->listed_in[i]);
    }
    free(s->listed_in);
}

Show procurar(Show array[], char *id){
    Show p;
    id[strcspn(id, "\n")] = '\0';
    for(int i = 0; i < 803; i++){
        if(strcmp(array[i].show_id, id) == 0){
            p = array[i];
            i = 803;
        }
    }
    return p;
}

void inserirArvore(ArvoreAVL *AVL, Show array[]){
    char id[40];
    scanf(" %[^\n\r]", id);
    while(IsFim(id)){
        Show x = procurar(array, id);
        inserir(AVL, x);
        scanf(" %[^\n\r]", id);
    }
}

void procurarNaArvore(ArvoreAVL *AVL){
    char nome[100];
    inicio = clock();
    scanf(" %[^\n\r]", nome);
    while(IsFim(nome) != 0){
        pesquisar(AVL, nome); 
        scanf(" %[^\n\r]", nome);
    }
    fim = clock();
    tempo = ((double)(fim - inicio)) / CLOCKS_PER_SEC;
}

void registroLog(){
    FILE *arquivo;
    arquivo = fopen("821811_avl.txt", "w");
    fprintf(arquivo, "Matrícula: 821811\t Tempo de execução: %.6f segundos\t Número de comparações: %d", tempo, count);
    fclose(arquivo);
}

int main(){
    Show shows[1368];
    char line[5000];
    int show_count = 0;
    ArvoreAVL *AVL = arvoreConstrutor();

    FILE *file = fopen("C:/Users/felip_000/Desktop/AED_2/2025_01/trabalhos_praticos/TP_04/disneyplus.csv", "r");
    if(!file){
        printf("Arquivo nao encontrado!");
        return 1;
    }

    fgets(line, sizeof(line), file);

    while(fgets(line, sizeof(line), file) && show_count < 1368){
        Show_init(&shows[show_count]);
        Show_Ler(&shows[show_count], line);
        show_count++;
    }
    fclose(file);

    inserirArvore(AVL, shows);
    procurarNaArvore(AVL);
    registroLog();

    for(int i = 0; i < show_count; i++){
        FreeShow(&shows[i]);
    }

    return 0;
}

File: Algoritimo-Estrutruras-de-Dados/Aeds-ll/2025_01/trabalhos_praticos/TP_04/questao_04/Show.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.time.Duration;
import java.time.Instant;
import java.util.*;

class NoAN{
    public boolean cor;
    public Show elemento;
    public NoAN esq, dir;

    public NoAN(){
        this(null, false, null, null);
    }

    public NoAN(Show elemento){
        this(elemento, false, null, null);
    }

    public NoAN(Show elemento, boolean cor){
        this(elemento, cor, null, null);
    }

    public NoAN(Show elemento, boolean cor, NoAN esq, NoAN dir){
        this.cor = cor;
        this.elemento = elemento;
        this.esq = esq;
        this.dir = dir;
    }
}

class Alvinegra{
    private NoAN raiz; 

    public Alvinegra(){
        raiz = null;
    }

    public void pesquisar(String elemento, Log log){
        System.out.print("=>raiz");
        boolean resp = pesquisar(elemento, raiz, log);

        if(resp){ 
            System.out.print(" SIM\n"); 
        } else{ 
            System.out.print(" NAO\n"); 
        }
    }

    private boolean pesquisar(String elemento, NoAN i, Log log){
        boolean resp;

        if(i == null){
            log.incrementaCount();
            resp = false;
        } else if(elemento.equals(i.elemento.getTitle())){
            log.incrementaCount();
            resp = true;
        } else if(elemento.compareTo(i.elemento.getTitle()) < 0){
            log.incrementaCount();
            System.out.print(" esq");
            resp = pesquisar(elemento, i.esq, log);
        } else{
            System.out.print(" dir");
            resp = pesquisar(elemento, i.dir, log);
        }

        return resp;
    }

    public void caminharCentral(){
        System.out.print("[ ");
        caminharCentral(raiz);
        System.out.println("]");
    }

    private void caminharCentral(NoAN i){
        if(i != null){
            caminharCentral(i.esq); 
            System.out.print(i.elemento + ((i.cor) ? "(p) " : "(b) ")); 
            caminharCentral(i.dir);
        }
    }

    public void caminharPre(){
        System.out.print("[ ");
        caminharPre(raiz);
        System.out.println("]");
    }

    private void caminharPre(NoAN i){
        if(i != null){
            System.out.print(i.elemento + ((i.cor) ? "(p) " : "(b) ")); 
            caminharPre(i.esq);
            caminharPre(i.dir);
        }
    }

    public void caminharPos(){
        System.out.print("[ ");
        caminharPos(raiz);
        System.out.println("]");
    }

    private void caminharPos(NoAN i){
        if(i != null){
            caminharPos(i.esq); 
            caminharPos(i.dir); 
            System.out.print(i.elemento + ((i.cor) ? "(p) " : "(b) ")); 
        }
    }

    public void inserir(Show elemento) throws Exception{
        if(raiz == null){
            raiz = new NoAN(elemento);
        } else if(raiz.esq == null && raiz.dir == null){
            if(elemento.getTitle().compareTo(raiz.elemento.getTitle()) < 0){
                raiz.esq = new NoAN(elemento);
            } else{
                raiz.dir = new NoAN(elemento);
            }
        } else if(raiz.esq == null){
            if(elemento.getTitle().compareTo(raiz.elemento.getTitle()) < 0){
                raiz.esq = new NoAN(elemento);
            } else if(elemento.getTitle().compareTo(raiz.dir.elemento.getTitle()) < 0){
                raiz.esq = new NoAN(raiz.elemento);
                raiz.elemento = elemento;
            } else{
                raiz.esq = new NoAN(raiz.elemento);
                raiz.elemento = raiz.dir.elemento;
                raiz.dir.elemento = elemento;
            }
            raiz.esq.cor = raiz.dir.cor = false;
        } else if(raiz.dir == null){
            if(elemento.getTitle().compareTo(raiz.elemento.getTitle()) > 0){
                raiz.dir = new NoAN(elemento);
            } else if(elemento.getTitle().compareTo(raiz.esq.elemento.getTitle()) > 0){
                raiz.dir = new NoAN(raiz.elemento);
                raiz.elemento = elemento;
            } else{
                raiz.dir = new NoAN(raiz.elemento);
                raiz.elemento = raiz.esq.elemento;
                raiz.esq.elemento = elemento;
            }
            raiz.esq.cor = raiz.dir.cor = false;

        } else{
            inserir(elemento, null, null, null, raiz);
        }
        raiz.cor = false;
    }

    private void balancear(NoAN bisavo, NoAN avo, NoAN pai, NoAN i){
        if(pai.cor == true){
            if(pai.elemento.getTitle().compareTo(avo.elemento.getTitle()) > 0){ 
                if(i.elemento.getTitle().compareTo(pai.elemento.getTitle()) > 0){
                    avo = rotacaoEsq(avo);
                } else{
                    avo = rotacaoDirEsq(avo);
                }
            } else{ 
                if(i.elemento.getTitle().compareTo(pai.elemento.getTitle()) < 0){
                    avo = rotacaoDir(avo);
                } else{
                    avo = rotacaoEsqDir(avo);
                }
            }

            if(bisavo == null){
                raiz = avo;
            } else if(avo.elemento.getTitle().compareTo(bisavo.elemento.getTitle()) < 0){
                bisavo.esq = avo;
            } else{
                bisavo.dir = avo;
            }
            avo.cor = false;
            avo.esq.cor = avo.dir.cor = true;
        } 
    }

    private void inserir(Show elemento, NoAN bisavo, NoAN avo, NoAN pai, NoAN i) throws Exception{
        if(i == null){
            if(elemento.getTitle().compareTo(pai.elemento.getTitle()) < 0){
                i = pai.esq = new NoAN(elemento, true);
            } else{
                i = pai.dir = new NoAN(elemento, true);
            }

            if(pai.cor == true){
                balancear(bisavo, avo, pai, i);
            }
        } else{
            if(i.esq != null && i.dir != null && i.esq.cor == true && i.dir.cor == true){
                i.cor = true;
                i.esq.cor = i.dir.cor = false;

                if(i == raiz){
                    i.cor = false;
                } else if(pai.cor == true){
                    balancear(bisavo, avo, pai, i);
                }
            }

            if(elemento.getTitle().compareTo(i.elemento.getTitle()) < 0){
                inserir(elemento, avo, pai, i, i.esq);
            } else if(elemento.getTitle().compareTo(i.elemento.getTitle()) > 0){
                inserir(elemento, avo, pai, i, i.dir);
            } else{
                throw new Exception("Erro inserir (elemento repetido)!");
            }
        }
    }

    private NoAN rotacaoDir(NoAN no){
        NoAN noEsq = no.esq;
        NoAN noEsqDir = noEsq.dir;

        noEsq.dir = no;
        no.esq = noEsqDir;

        return noEsq;
    }

    private NoAN rotacaoEsq(NoAN no){
        NoAN noDir = no.dir;
        NoAN noDirEsq = noDir.esq;

        noDir.esq = no;
        no.dir = noDirEsq;

        return noDir;
    }

    private NoAN rotacaoDirEsq(NoAN no){
        no.dir = rotacaoDir(no.dir);

        return rotacaoEsq(no);
    }

    private NoAN rotacaoEsqDir(NoAN no){
        no.esq = rotacaoEsq(no.esq);

        return rotacaoDir(no);
    }
}

class Log{
    public int count;
    public float time;
    public String fileName;
     
    Log(String fileName){
        this.count = 0;
        this.fileName = fileName;
    }
 
    public void incrementaCount(){
        count++;
    }
 
    public void registrarLog(){
        try{
            FileWriter writer = new FileWriter(this.fileName);
            writer.write("Matrícula: 821811 \ttempo de execução: " + this.time + " segundos  \tNúmero de comparações: " + this.count);
            writer.close();

        }catch(IOException e){
            System.out.println("Erro: " + e.getMessage());
        }
    }
}

class Data{
    private String mes;
    private int dia;
    private int ano;

    public Data(){
        this.mes = "NaN";
        this.dia = 0;
        this.ano = 0;
    }

    public Data(String mes, int dia, int ano){
        this.mes = mes;
        this.dia = dia;
        this.ano = ano;
    }

    public String getMes(){
        return mes;
    }
    public void setMes(String mes){
        this.mes = mes;
    }

    public int getDia(){
        return dia;
    }
    public void setDia(int dia){
        this.dia = dia;
    }

    public int getAno(){
        return ano;
    }
    public void setAno(int ano){
        this.ano = ano;
    }

    public String toString() {
        return mes + " " + dia + ", " + ano;
    }

    public static Data Converter(String data){
        String mes = "NaN";
        int dia = 0;
        int ano = 0;
    
        if(!data.equals("NaN")){
            String[] partes = data.split(" ");
            mes = partes[0];
            dia = Integer.parseInt(partes[1].replace(",", ""));
            ano = Integer.parseInt(partes[2]);
        }
    
        return new Data(mes, dia, ano);
    }    
}

class Show{
    private String show_id;
    private String type;
    private String title;
    private String director;
    private String[] cast;
    private String country;
    private Data data;
    private String release_year;
    private String rating;
    private String duration;
    private String[] listed_in;

    public Show(){
        this.show_id = "";
        this.type = "";
        this.title = "";
        this.director = "";
        this.cast = new String[]{"NaN"};
        this.country = "";
        this.data = new Data();
        this.release_year = "";
        this.rating = "";
        this.duration = "";
        this.listed_in = new String[]{"NaN"};
    }

    public String getShow_id(){
        return show_id;
    }
    public void setShow_id(String show_id){
        this.show_id = show_id;
    }

    public String getType(){
        return type;
    }
    public void setType(String type){
        this.type = type;
    }

    public String getTitle(){
        return title;
    }
    public void setTitle(String title){
        this.title = title;
    }

    public String getDirector(){
        return director;
    }
    public void setDirector(String director){
        this.director = director;
    }

    public String[] getCast(){
        return cast;
    }
    public void setCast(String[] cast){
        this.cast = cast;
    }

    public String getCountry(){
        return country;
    }
    public void setCountry(String country){
        this.country = country;
    }

    public Data getData(){
        return data;
    }
    public void setData(Data data){
        this.data = data;
    }

    public String getRelease_year(){
        return release_year;
    }
    public void setRelease_year(String release_year){
        this.release_year = release_year;
    }

    public String getRating(){
        return rating;
    }
    public void setRating(String rating){
        this.rating = rating;
    }

    public String getDuration(){
        return duration;
    }
    public void setDuration(String duration){
        this.duration = duration;
    }

    public String[] getListed_in(){
        return listed_in;
    }
    public void setListed_in(String[] listed_in){
        this.listed_in = listed_in;
    }

    public Show clone(){
    String[] castClone = (this.cast != null) ? this.cast.clone() : null;
    String[] listedInClone = (this.listed_in != null) ? this.listed_in.clone() : null;
    String dataString = (this.data != null) ? this.data.toString() : null;

    return new Show(this.show_id, this.type, this.title, this.director,
                    castClone, this.country, dataString, this.release_year,
                    this.rating, this.duration, listedInClone);
    }

    public Show(String show_id, String type, String title, String director, String[] cast, String country, String date_added,
                String release_year, String rating, String duration, String[] listed_in){
        this.show_id = show_id;
        this.type = type;
        this.title = title;
        this.director = director;
        this.cast = cast;
        this.country = country;
        this.data = Data.Converter(date_added);
        this.release_year = release_year;
        this.rating = rating;
        this.duration = duration;
        this.listed_in = listed_in;
    }

    public static void ordenar(String[] array){
        for(int i = 0; i < array.length - 1; i++){
            for(int j = 0; j < array.length - i - 1; j++){
                if(array[j].compareTo(array[j + 1]) > 0){
                    String temp = array[j];
                    array[j] = array[j + 1];
                    array[j + 1] = temp;
                }
            }
        }
    }    

    public static Show Ler(String line){
        Show s = new Show();
        String[] temp = new String[12];
        int index = 0;
        boolean dentroAspas = false;
        StringBuilder campo = new StringBuilder();
    
        for(int i = 0; i < line.length(); i++){
            char c = line.charAt(i);
    
            if(c == '\"'){
                dentroAspas = !dentroAspas;
            } else if(c == ',' && !dentroAspas){
                temp[index++] = campo.length() == 0 ? "NaN" : campo.toString();
                campo.setLength(0); 
            } else{
                campo.append(c);
            }
        }
    
        temp[index] = campo.length() == 0 ? "NaN" : campo.toString();
    
        s.show_id      = temp[0];
        s.type         = temp[1];
        s.title        = temp[2];
        s.director     = temp[3];
        s.cast = temp[4].equals("NaN") ? new String[]{"NaN"} : temp[4].split(",\\s*");
        if(!temp[4].equals("NaN")){
            ordenar(s.cast);
        }
        s.country      = temp[5];
        s.data         = Data.Converter(temp[6]);
        s.release_year = temp[7];
        s.rating       = temp[8];
        s.duration     = temp[9];
        s.listed_in = temp[10].equals("NaN") ? new String[]{"NaN"} : temp[10].split(",\\s*");
        if(!temp[10].equals("NaN")){
            ordenar(s.listed_in);
        }
    
        return s;
    }
     
    public void Imprimir(){
        System.out.print("=> " + show_id + " ## ");
        System.out.print(title + " ## ");
        System.out.print(type + " ## ");
        System.out.print(director + " ## ");
    
        System.out.print("[");
        if(cast.length == 1 && cast[0].equals("NaN")){
            System.out.print("NaN");
        } else {
            for(int i = 0; i < cast.length; i++){
                System.out.print(cast[i]);
                if(i < cast.length - 1) System.out.print(", ");
            }
        }
        System.out.print("] ## ");
    
        System.out.print(country + " ## ");
        System.out.print(data.toString() + " ## ");
        System.out.print(release_year + " ## ");
        System.out.print(rating + " ## ");
        System.out.print(duration + " ## ");
    
        System.out.print("[");
        if(listed_in.length == 1 && listed_in[0].equals("NaN")){
            System.out.print("NaN");
        } else {
            for(int i = 0; i < listed_in.length; i++){
                System.out.print(listed_in[i]);
                if(i < listed_in.length - 1) System.out.print(", ");
            }
        }
        System.out.println("] ## ");
    }       

    public static void Carregar(Show[] shows){
        File file = new File("/tmp/disneyplus.csv");

        try{
            Scanner sc = new Scanner(file);
            sc.nextLine();
    
            for(int i = 0; i < 1368; i++){
                String linha = sc.nextLine();
                shows[i] = Show.Ler(linha);
            }
    
            sc.close();
        } catch(FileNotFoundException e){
            System.out.println("Arquivo não encontrado!");
        }
    }

    public static int IsFim(String line){
        int result = 1; 

        if(line.charAt(0) == 'F' && line.charAt(1) == 'I' && line.charAt(2) == 'M'){
            result = 0;
        }

        return result;
    }

    public static void swap(Show s[], int i, int j){
        Show tmp = s[i];
        s[i] = s[j];
        s[j] = tmp;
    }

    public static Show Procurar(Show s[], String id){
        Show x = new Show();

        for(int i = 0; i < s.length; i++){
            if(s[i].getShow_id().equals(id)){
                x = s[i];
                i = s.length;
            }
        }
        return x;
    }

    public static void inserirArvore(Alvinegra arvore, Show shows[], Scanner sc) throws Exception{
        String id;

        id = sc.nextLine();
        while(IsFim(id) != 0){
            Show novo;
            novo = Procurar(shows, id);
            arvore.inserir(novo);
            id = sc.nextLine();
        }
    }

    public static void procuraNaArvore(Alvinegra arvore, Log log,Scanner sc){
        String nome; 

        nome = sc.nextLine();
        while(IsFim(nome) != 0){ 
            arvore.pesquisar(nome, log);
            nome = sc.nextLine();
        }
    }

    public static void main(String[] args) throws Exception{
        Scanner sc = new Scanner(System.in);
        Show[] shows = new Show[1368];
        Alvinegra arvore = new Alvinegra();
        Log log = new Log("/tmp/821811_arvoreAlvinegra.txt");

        Carregar(shows);
        inserirArvore(arvore, shows, sc);
        Instant start = Instant.now();
        procuraNaArvore(arvore, log, sc);
        Instant end = Instant.now();
        long elapsedTime = Duration.between(start, end).toMillis();
        log.time = (float) elapsedTime / 1000; 
        log.registrarLog();
        sc.close();
    }
}